_posts/2015-09-12-developing-gamepad-virtual-keyboard.markdown

```diff
@@ -0,0 +1,150 @@
+
+The functionality for those last 3 points all fold into each other, so we’ll discuss them all at once.
+
+Let’s assume you already have a mechanism for getting the currently selected symbol. We'll need a callback for that event, let's call it onGamepadEvent. Inside this callback we receive the new character being typed out. We can't just concatenate it onto the end of our current string because the cursor could be anywhere. So we need to get the current string value from the input, get the cursor location within that string, and then insert the new character at that location of the cursor. We also need to increment the location of the cursor, as I'll discuss next.
+
+```js
+// stores/input.js
+
+getValueDivision: function() {
+    var value = this.value + '';
+    var firstStrPart = value.substring(0, this.cursor),
+        secondStrPart = value.substring(this.cursor, value.length);
+
+    return {
+        start: firstStrPart,
+        end: secondStrPart
+    }
+},
+
+onGamepadEvent: function(symbol) {
+    var valueParts = this.getValueDivision();
+    this.value = valueParts.start + symbol + valueParts.end;
+    this.cursor++;
+    this.emit('change');
+},
+
+```
+
+Since the location of the cursor can be changed by the mouse, keyboard, and the gamepad it's impractical to listen for keyboard events and mouse clicks to track the cursor, and in turn, know where to add new characters. Instead, let's use the onChange event for the Fake input and update a value that we store to keep track of the cursor. The DOM provides a decent API to work with text selection and the cursor, called the [Selection API](https://developer.mozilla.org/en-US/docs/Web/API/Selection). You'll need polyfills to get support for IE, here are the ones for [getCursor](http://stackoverflow.com/a/263796/2687479) and [setCursor](http://stackoverflow.com/questions/1865563/set-cursor-at-a-length-of-14-onfocus-of-a-textbox/1867393#1867393)
+
+```js
+// jsx/InputView.js
+
+onInputChange: function() {
+    var inputEl = this.refs.input.getDOMNode();
+    var value = inputEl.value;
+    flux.actions.setInputValue(value);
+    this.onInputMouseDown();
+},
+
+onInputMouseDown: function() {
+    var inputEl = this.refs.input.getDOMNode();
+    var cursorLocation = Utils.getCursor(inputEl);
+    flux.actions.setInputCursor(cursorLocation);
+},
+
+```
+
+Whenever the input is changed we get the new location of the cursor and then manually set it there. This way we can use that same system to connect the DPad of the gamepad to the cursor location as well. Again, if you're using Flux, this is easy to manage using a store for the cursor's location, then reflecting that value to anything that needs it.
```

Comment:
Again, if you're using Flux, this is easy to manage using a store for the cursor's location, which in turn reflects that value to anything that needs it.